src/pool.rs

```diff
@@ -225,13 +225,13 @@ impl PoolKey {
 /// *Internal API*
 pub(crate) struct PoolReturnRead<R: Read + Sized + Into<Stream>> {
     // unit that contains the agent where we want to return the reader.
-    unit: Option<Unit>,
+    unit: Box<Option<Unit>>,
```

Comment:
I think we want this the other way around:

```suggestion
    unit: Option<Box<Unit>>,
```

Because of the null pointer optimization. I wasn't able to find documentation dedicated to the NPO specifically but it's mentioned several places, including here: https://doc.rust-lang.org/nomicon/repr-rust.html

> The classic case of this is Rust's "null pointer optimization": an enum consisting of a single outer unit variant (e.g. None) and a (potentially nested) non- nullable pointer variant (e.g. Some(&T)) makes the tag unnecessary. A null pointer can safely be interpreted as the unit (None) variant. The net result is that, for example, size_of::<Option<&T>>() == size_of::<&T>().

Reply:
done